File: prototype/php/src/app/Actions/Orders/RollUpOrderStatus.php
<?php

declare(strict_types=1);

namespace App\Actions\Orders;

use App\Domain\Orders\OrderStatus;
use App\Models\Fulfillment;
use App\Models\Order;

final class RollUpOrderStatus
{
    public function __invoke(Order $order): Order
    {
        $order->update([
            'status' => OrderStatus::fromFulfillments(
                array_values($order->fulfillments()->get()->map(fn (Fulfillment $fulfillment) => $fulfillment->status)->all()),
            ),
        ]);

        return $order;
    }
}

File: prototype/rails/src/app/actions/orders/roll_up_order_status.rb
module Orders
  class RollUpOrderStatus
    def call(order:)
      # Reloaded because the caller reached the order through a fulfillment it
      # has already changed, and the cached collection still holds the old row.
      order.update!(status: Domain::Orders::OrderStatus.from_fulfillments(order.fulfillments.reload.pluck(:status)))

      order
    end
  end
end
